File: 13-02-2024_Clone_an_undirected_graph.cpp
class Solution {
public:
    Node* cloneGraph(Node* node) {
        map<int, Node*> nodeMap;
        return cloneHelper(node, nodeMap);
    }

    Node* cloneHelper(Node* node, map<int, Node*>& nodeMap) {
        if (node == nullptr)
            return nullptr;

        if (nodeMap.find(node->val) == nodeMap.end()) {
            nodeMap[node->val] = new Node(node->val);
            for (Node* neighbor : node->neighbors) {
                if (nodeMap.find(neighbor->val) == nodeMap.end()) {
                    nodeMap[node->val]->neighbors.push_back(cloneHelper(neighbor, nodeMap));
                } else {
                    nodeMap[node->val]->neighbors.push_back(nodeMap[neighbor->val]);
                }
            }
        }
        return nodeMap[node->val];
    }
};

File: 13-02-2024_Clone_an_undirected_graph.java
class Solution {
    Node cloneGraph(Node node) {
        Map<Integer, Node> nodeMap = new HashMap<>();
        return cloneHelper(node, nodeMap);
    }

    Node cloneHelper(Node node, Map<Integer, Node> nodeMap) {
        if (node == null)
            return null;

        if (!nodeMap.containsKey(node.val)) {
            nodeMap.put(node.val, new Node(node.val));
            for (Node neighbor : node.neighbors) {
                if (!nodeMap.containsKey(neighbor.val)) {
                    nodeMap.get(node.val).neighbors.add(cloneHelper(neighbor, nodeMap));
                } else {
                    nodeMap.get(node.val).neighbors.add(nodeMap.get(neighbor.val));
                }
            }
        }
        return nodeMap.get(node.val);
    }

}
